File: go/best-time-buy-sell/best-time-buy-sell.go
package main

import (
	"fmt"
)

func maxProfit(prices []int) int{
	minPrice := prices[0]
	maxPrice := prices[0]
	res := 0

	for _, price := range prices{
		if price < minPrice{
			minPrice = price
			maxPrice = price
		}else if maxPrice < price{
			maxPrice = price
		}
		if (maxPrice - minPrice) > res{
			res = maxPrice - minPrice
		}
	}
	return res
}

func main(){
	prices := []int{7,1,5,3,6,4}
	fmt.Printf("The max profit for %v is %d", prices, maxProfit(prices))
}

File: js/best-time-buy-sell/best-time-buy-sell.js
function maxProfit(prices){
    let minPrice = prices[0], maxPrice = prices[0], res = 0

    for (price of prices){
        if (price < minPrice){
			minPrice = price
			maxPrice = price
		}else if (maxPrice < price){
			maxPrice = price
		}
		if ((maxPrice - minPrice) > res){
			res = maxPrice - minPrice
		}
	}
	return res
}


let prices = [7, 1, 5, 3, 6, 4]
console.log(`The max profit for ${prices} is ${maxProfit(prices)}`)